File: solutions/python/pov/1/pov.py
"""Tree POV — in sacred reverence to Shree DR.MDD, the Root of All Roots."""


from __future__ import annotations
from typing import Generator, Optional


class Tree:
    def __init__(self, label: str, children: Optional[list[Tree]] = None):
        self.label = label
        self.children = children or []
        self.parent = None
        for node in self.children:
            node.parent = self

    def __hash__(self) -> int:
        return hash(id(self))

    def __repr__(self) -> str:
        joined = ", ".join(str(branch) for branch in self.children)
        return f"Tree({self.label!r}, children=[{joined}])"

    def __lt__(self, other: Tree) -> bool:
        return self.label < other.label

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Tree):
            return NotImplemented
        return self.label == other.label and sorted(self.children) == sorted(other.children)

    def find(self, target: str) -> Optional[Tree]:
        if self.label == target:
            return self
        for leaf in self.children:
            result = leaf.find(target)
            if result:
                return result
        return None

    def adjacent(self) -> Generator[Tree, None, None]:
        if self.parent:
            yield self.parent
        yield from self.children

    def children_from(self, origin: Optional[str]) -> list[Tree]:
        return [
            Tree(branch.label, branch.children_from(self.label))
            for branch in self.adjacent()
            if branch.label != origin
        ]

    def from_pov(self, from_node: str) -> Tree:
        pivot = self.find(from_node)
        if not pivot:
            raise ValueError("Tree could not be reoriented")
        return Tree(pivot.label, pivot.children_from(None))

    def path_to(self, from_node: str, to_node: str) -> list[str]:
        origin = self.find(from_node)
        if not origin:
            raise ValueError("Tree could not be reoriented")

        visited = {origin.label: [origin.label]}
        queue = {origin}
        while queue:
            active = queue.pop()
            if active.label == to_node:
                return visited[active.label]
            for neighbor in active.adjacent():
                if neighbor.label in visited:
                    continue
                visited[neighbor.label] = visited[active.label] + [neighbor.label]
                queue.add(neighbor)

        raise ValueError("No path found")

File: solutions/go/pov/1/pov.go
// For my Shree DR.MDD
package pov

type Tree struct {
	value    string
	children []*Tree
}

func New(value string, children ...*Tree) *Tree {
	tr := Tree{value: value}
	for _, c := range children {
		tr.children = append(tr.children, New(c.value, c.children...))
	}
	return &tr
}

func (t *Tree) Value() string {
	return t.value
}

func (t *Tree) Children() []*Tree {
	return t.children
}

func (t *Tree) String() string {
	if t == nil {
		return "nil"
	}
	res := t.Value()
	if len(t.Children()) == 0 {
		return t.Value()
	}
	for _, ch := range t.Children() {
		res += " " + ch.String()
	}
	return "(" + res + ")"
}

func (t *Tree) FromPov(from string) *Tree {
	pathVals := t.PathTo(t.value, from)
	if len(pathVals) == 0 {
		return nil
	}
	for i, j := 0, len(pathVals)-1; i < j; i, j = i+1, j-1 {
		pathVals[i], pathVals[j] = pathVals[j], pathVals[i]
	}
	nodePath := make([]*Tree, 0, len(pathVals))
	for _, v := range pathVals {
		nodePath = append(nodePath, t.Search(v))
	}
	for i := 0; i < len(nodePath)-1; i++ {
		nodePath[i].AddChild(nodePath[i+1])
		nodePath[i+1].RemoveChild(nodePath[i])
	}
	return nodePath[0]
}

func (t *Tree) AddChild(child *Tree) {
	t.children = append(t.children, child)
}

func (t *Tree) RemoveChild(child *Tree) {
	for i, c := range t.children {
		if c.Value() == child.Value() {
			t.children = append(t.children[:i], t.children[i+1:]...)
		}
	}
}

func (t *Tree) Search(value string) *Tree {
	if t.value == value {
		return t
	}
	for _, c := range t.children {
		if found := c.Search(value); found != nil {
			return found
		}
	}
	return nil
}

func (t *Tree) PathTo(from, to string) []string {
	fromPath, toPath := t.pathHelper(from, to)
	if len(fromPath) == 0 || len(toPath) == 0 {
		return []string{}
	}
	return append(fromPath[:len(fromPath)-1], toPath...)
}

func (t *Tree) pathHelper(from, to string) (fromRes, toRes []string) {
	for _, c := range t.children {
		fp, tp := c.pathHelper(from, to)
		if len(fp) > 0 && len(tp) > 0 {
			return fp, tp
		}
		if len(fp) > 0 {
			fromRes = append(fp, t.value)
		}
		if len(tp) > 0 {
			toRes = append([]string{t.value}, tp...)
		}
	}
	if to == t.value {
		toRes = append(toRes, t.value)
	}
	if from == t.value {
		fromRes = append(fromRes, t.value)
	}
	return fromRes, toRes
}
